DOC/man/man3/Case.3
.TH "Case" 3 "Dimanche 16 Août 2020" "Demineur" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Case
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Case\&.h>\fP
.PP
Est dérivée de QPushButton\&.
.SS "Connecteurs publics"

.in +1c
.ti -1c
.RI "void \fBincrementNbDrapeauAutour\fP ()"
.br
.RI "Incréménte le nombre de drapeaux autour d'elle\&. "
.ti -1c
.RI "void \fBdecrementNbDrapeauAutour\fP ()"
.br
.RI "Décrémente le nombre de drapeaux autour d'elle\&. "
.ti -1c
.RI "void \fBincrementNbMinesAutour\fP ()"
.br
.RI "Incrémente le nomre de mine autour d'elle\&. "
.in -1c
.SS "Signaux"

.in +1c
.ti -1c
.RI "void \fBPremierClick\fP (\fBCase\fP *)"
.br
.ti -1c
.RI "void \fBClickNormal\fP ()"
.br
.ti -1c
.RI "void \fBClickDroit\fP ()"
.br
.ti -1c
.RI "void \fBClickGaucheEtDroit\fP ()"
.br
.ti -1c
.RI "void \fBPasDeClick\fP ()"
.br
.ti -1c
.RI "void \fBExplosion\fP ()"
.br
.ti -1c
.RI "void \fBreveal\fP ()"
.br
.ti -1c
.RI "void \fBrevealed\fP ()"
.br
.ti -1c
.RI "void \fBrevealNeighbors\fP ()"
.br
.ti -1c
.RI "void \fBpreview\fP ()"
.br
.ti -1c
.RI "void \fBunPreview\fP ()"
.br
.ti -1c
.RI "void \fBunPreviewNeighbors\fP ()"
.br
.ti -1c
.RI "void \fBPointDrapeau\fP (bool)"
.br
.ti -1c
.RI "void \fBSuppressionDrappeau\fP (bool)"
.br
.ti -1c
.RI "void \fBInactivation\fP ()"
.br
.in -1c
.SS "Fonctions membres publiques"

.in +1c
.ti -1c
.RI "\fBCase\fP (\fBPosition\fP \fBposition\fP, QWidget *parent=nullptr)"
.br
.ti -1c
.RI "virtual \fB~Case\fP () override"
.br
.ti -1c
.RI "void \fBAfficheResultat\fP ()"
.br
.RI "Corrige les Affichages à la fin de la partie\&. "
.ti -1c
.RI "void \fBAjoutVoisin\fP (\fBCase\fP *\fBCase\fP)"
.br
.ti -1c
.RI "\fBPosition\fP \fBposition\fP () const"
.br
.ti -1c
.RI "void \fBplacerMine\fP (bool val)"
.br
.ti -1c
.RI "bool \fBEstMine\fP () const"
.br
.RI "Informe sur l'existence ou non d'une Mine\&. "
.ti -1c
.RI "bool \fBAuneMarque\fP () const"
.br
.RI "Informe sur l'existence ou non d'un marquage de drapeau\&. "
.ti -1c
.RI "bool \fBEstOuvert\fP () const"
.br
.ti -1c
.RI "bool \fBisUnrevealed\fP () const"
.br
.ti -1c
.RI "bool \fBAdesMinesAutour\fP () const"
.br
.RI "Informe sur l'existence de mines dans au moins l'une des cases voisines\&. "
.ti -1c
.RI "unsigned int \fBNbMinesAutour\fP () const"
.br
.RI "Informe sur le nombre de mines dans les cases voisines\&. "
.ti -1c
.RI "unsigned int \fBNbDrapeauAutour\fP () const"
.br
.RI "Retourne le nombre de drapeaux autour d'elle A la fin du jeux si chaque case a le même nombre de drapeaux autour égal au nombre de mine autour c'est que les mines ont bien été ifentifiées donc le jeu peut s'arrêter\&. "
.ti -1c
.RI "QList< \fBCase\fP * > & \fBVoisinage\fP ()"
.br
.RI "Liste des cases qui contituent le voisinage\&. "
.ti -1c
.RI "virtual void \fBmousePressEvent\fP (QMouseEvent *e) override"
.br
.ti -1c
.RI "virtual void \fBmouseReleaseEvent\fP (QMouseEvent *e) override"
.br
.ti -1c
.RI "virtual void \fBmouseMoveEvent\fP (QMouseEvent *e) override"
.br
.ti -1c
.RI "virtual QSize \fBsizeHint\fP () const override"
.br
.in -1c
.SH "Description détaillée"
.PP 
Définition à la ligne 15 du fichier Case\&.h\&.
.SH "Documentation des constructeurs et destructeur"
.PP 
.SS "Case::Case (\fBPosition\fP position, QWidget * parent = \fCnullptr\fP)"

.PP
Définition à la ligne 16 du fichier Case\&.cpp\&.
.SS "Case::~Case ()\fC [override]\fP, \fC [virtual]\fP"

.PP
Définition à la ligne 29 du fichier Case\&.cpp\&.
.SH "Documentation des fonctions membres"
.PP 
.SS "bool Case::AdesMinesAutour () const"

.PP
Informe sur l'existence de mines dans au moins l'une des cases voisines\&. Si une case voisine est minée elle retourne Vrai 
.PP
Définition à la ligne 71 du fichier Case\&.cpp\&.
.SS "void Case::AfficheResultat ()"

.PP
Corrige les Affichages à la fin de la partie\&. 
.PP
Définition à la ligne 349 du fichier Case\&.cpp\&.
.SS "void Case::AjoutVoisin (\fBCase\fP * Case)"

.PP
Définition à la ligne 39 du fichier Case\&.cpp\&.
.SS "bool Case::AuneMarque () const"

.PP
Informe sur l'existence ou non d'un marquage de drapeau\&. 
.PP
Définition à la ligne 94 du fichier Case\&.cpp\&.
.SS "void Case::ClickDroit ()\fC [signal]\fP"

.SS "void Case::ClickGaucheEtDroit ()\fC [signal]\fP"

.SS "void Case::ClickNormal ()\fC [signal]\fP"

.SS "void Case::decrementNbDrapeauAutour ()\fC [slot]\fP"

.PP
Décrémente le nombre de drapeaux autour d'elle\&. 
.PP
Définition à la ligne 127 du fichier Case\&.cpp\&.
.SS "bool Case::EstMine () const"

.PP
Informe sur l'existence ou non d'une Mine\&. Si la case est minée elle retourne Vrai 
.PP
Définition à la ligne 62 du fichier Case\&.cpp\&.
.SS "bool Case::EstOuvert () const"

.PP
Définition à la ligne 99 du fichier Case\&.cpp\&.
.SS "void Case::Explosion ()\fC [signal]\fP"

.SS "void Case::Inactivation ()\fC [signal]\fP"

.SS "void Case::incrementNbDrapeauAutour ()\fC [slot]\fP"

.PP
Incréménte le nombre de drapeaux autour d'elle\&. 
.PP
Définition à la ligne 120 du fichier Case\&.cpp\&.
.SS "void Case::incrementNbMinesAutour ()\fC [slot]\fP"

.PP
Incrémente le nomre de mine autour d'elle\&. 
.PP
Définition à la ligne 87 du fichier Case\&.cpp\&.
.SS "bool Case::isUnrevealed () const"

.PP
Définition à la ligne 104 du fichier Case\&.cpp\&.
.SS "void Case::mouseMoveEvent (QMouseEvent * e)\fC [override]\fP, \fC [virtual]\fP"

.PP
Définition à la ligne 170 du fichier Case\&.cpp\&.
.SS "void Case::mousePressEvent (QMouseEvent * e)\fC [override]\fP, \fC [virtual]\fP"

.PP
Définition à la ligne 139 du fichier Case\&.cpp\&.
.SS "void Case::mouseReleaseEvent (QMouseEvent * e)\fC [override]\fP, \fC [virtual]\fP"

.PP
Définition à la ligne 160 du fichier Case\&.cpp\&.
.SS "unsigned int Case::NbDrapeauAutour () const"

.PP
Retourne le nombre de drapeaux autour d'elle A la fin du jeux si chaque case a le même nombre de drapeaux autour égal au nombre de mine autour c'est que les mines ont bien été ifentifiées donc le jeu peut s'arrêter\&. 
.PP
Définition à la ligne 113 du fichier Case\&.cpp\&.
.SS "unsigned int Case::NbMinesAutour () const"

.PP
Informe sur le nombre de mines dans les cases voisines\&. Elle retour un Entier Positif 
.PP
Définition à la ligne 80 du fichier Case\&.cpp\&.
.SS "void Case::PasDeClick ()\fC [signal]\fP"

.SS "void Case::placerMine (bool val)"

.PP
Définition à la ligne 51 du fichier Case\&.cpp\&.
.SS "void Case::PointDrapeau (bool)\fC [signal]\fP"

.SS "\fBPosition\fP Case::position () const"

.PP
Définition à la ligne 46 du fichier Case\&.cpp\&.
.SS "void Case::PremierClick (\fBCase\fP *)\fC [signal]\fP"

.SS "void Case::preview ()\fC [signal]\fP"

.SS "void Case::reveal ()\fC [signal]\fP"

.SS "void Case::revealed ()\fC [signal]\fP"

.SS "void Case::revealNeighbors ()\fC [signal]\fP"

.SS "QSize Case::sizeHint () const\fC [override]\fP, \fC [virtual]\fP"

.PP
Définition à la ligne 179 du fichier Case\&.cpp\&.
.SS "void Case::SuppressionDrappeau (bool)\fC [signal]\fP"

.SS "void Case::unPreview ()\fC [signal]\fP"

.SS "void Case::unPreviewNeighbors ()\fC [signal]\fP"

.SS "QList< \fBCase\fP * > & Case::Voisinage ()"

.PP
Liste des cases qui contituent le voisinage\&. 
.PP
Définition à la ligne 134 du fichier Case\&.cpp\&.

.SH "Auteur"
.PP 
Généré automatiquement par Doxygen pour Demineur à partir du code source\&.


Case.cpp
#include "Case.h"
#include <thread>
#include <chrono>
#include "Styles.h"
#include <QDebug>
#include <QState>
#include <QFinalState>
#include <QMouseEvent>
#include <QSizePolicy>

bool Case::_firstClick = false;
QSize Case::Dimension= QSize(50,50);



Case::Case(Position position, QWidget* parent /*= nullptr*/)
    : _isMine(false)
    , _NbMinesAutour(0)
    , _NbDrapeauAutour(0)
    , _position(position)
    , QPushButton(parent)
{
    ConfigGestionEtatCase();
    this->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
    setCheckable(true);
    setMouseTracking(true);
}

Case::~Case()
{
    _firstClick = false;
    delete ModeNormal;
    delete ModeAppercu;
    delete ModeDrapeauPointe;
    delete ModeDecouvert;
    delete ModeInactive;
}

void Case::AjoutVoisin(Case* Case)
{
    _voisins += Case;
    connect(this, &Case::revealNeighbors, Case, &Case::reveal, Qt::QueuedConnection);
    connect(this, &Case::unPreviewNeighbors, Case, &Case::unPreview, Qt::QueuedConnection);
}

Position Case::position() const
{
    return _position;
}

void Case::placerMine(bool val)
{
    _isMine = val;
    for (auto voisin : _voisins)
        voisin->incrementNbMinesAutour();
}
/*!
    *  \brief Informe sur l'existence ou non d'une Mine
    *
    *  Si la case est minée elle retourne Vrai
    */
bool Case::EstMine() const
{
    return _isMine;
}
/*!
    *  \brief Informe sur l'existence de mines dans au moins l'une des cases voisines
    *
    *  Si une case voisine est minée elle retourne Vrai
    */
bool Case::AdesMinesAutour() const
{
    return _NbMinesAutour;
}
/*!
    *  \brief Informe sur le nombre de mines dans les cases voisines
    *
    *  Elle retour un Entier Positif
    */
unsigned int Case::NbMinesAutour() const
{
    return _NbMinesAutour;
}
/*!
    *  \brief Incrémente le nomre de mine autour d'elle
    */
void Case::incrementNbMinesAutour()
{
    ++_NbMinesAutour;
}
/*!
    *  \brief Informe sur l'existence ou non d'un marquage de drapeau
    */
bool Case::AuneMarque() const
{
    return _machine.configuration().contains(ModeDrapeauPointe);
}

bool Case::EstOuvert() const
{
    return _machine.configuration().contains(ModeDecouvert);
}

bool Case::isUnrevealed() const
{
    return _machine.configuration().contains(ModeNormal);
}
/*!
    *  \brief Retourne le nombre de drapeaux autour d'elle
    * A la fin du jeux si chaque case a le même nombre de drapeaux autour égal au  nombre de mine autour c'est
    * que les mines ont bien été ifentifiées donc le jeu peut s'arrêter
    */
unsigned int Case::NbDrapeauAutour() const
{
    return _NbDrapeauAutour;
}
/*!
    *  \brief Incréménte le nombre de drapeaux autour d'elle
    */
void Case::incrementNbDrapeauAutour()
{
    ++_NbDrapeauAutour;
}
/*!
    *  \brief Décrémente le nombre de drapeaux autour d'elle
    */
void Case::decrementNbDrapeauAutour()
{
    --_NbDrapeauAutour;
}
/*!
    *  \brief Liste des cases qui contituent le voisinage
    */
QList<Case*>& Case::Voisinage()
{
    return _voisins;
}

void Case::mousePressEvent(QMouseEvent* e)
{
    if (!_firstClick)
    {
        _firstClick = true;
        emit PremierClick(this);
    }

    _bothClicked = false;

    if (e->buttons() == (Qt::LeftButton | Qt::RightButton))
    {
        emit ClickGaucheEtDroit();
        _buttons = Qt::LeftButton | Qt::RightButton;
    }
    else if (e->buttons() == Qt::LeftButton)
        _buttons = Qt::LeftButton;
    else if (e->buttons() == Qt::RightButton)
        _buttons = Qt::RightButton;
}

void Case::mouseReleaseEvent(QMouseEvent* e)
{
    if (_buttons == (Qt::LeftButton | Qt::RightButton))
        emit PasDeClick();
    else if (_buttons == Qt::LeftButton)
        emit ClickNormal();
    else if (_buttons == Qt::RightButton)
        emit ClickDroit();
}

void Case::mouseMoveEvent(QMouseEvent* e)
{
    if (e->buttons() == (Qt::LeftButton | Qt::RightButton))
    {
        if (!this->rect().contains(this->mapFromGlobal(QCursor::pos())))
            emit unPreview();
    }
}

QSize Case::sizeHint() const
{
    return Dimension;
}
/*!
    *  \brief Configuration de la Machine des Etats d'une case
    *
    *  Configuration de la machine qui suit les Etats et les transitions possible sur
    * une Case

    */

void Case::ConfigGestionEtatCase()
{
    ModeNormal = new QState;
    ModeAppercu = new QState;
    AppercuDesVoisins = new QState;
    ModeDrapeauPointe = new QState;
    ModeDecouvert = new QState;
    ModeVoisinsDecouverts = new QState;
    ModeInactive = new QFinalState;
// Gestion des transtions
    // quitter Etat normal de la case aux evenements
    ModeNormal->addTransition(this, &Case::ClickNormal, ModeDecouvert);
    ModeNormal->addTransition(this, &Case::ClickDroit, ModeDrapeauPointe);
    ModeNormal->addTransition(this, &Case::reveal, ModeDecouvert);
    ModeNormal->addTransition(this, &Case::preview, ModeAppercu);
    ModeNormal->addTransition(this, &Case::Inactivation, ModeInactive);

    ModeAppercu->addTransition(this, &Case::reveal, ModeDecouvert);
    ModeAppercu->addTransition(this, &Case::unPreview, ModeNormal);
    ModeAppercu->addTransition(this, &Case::Inactivation, ModeInactive);

    ModeDrapeauPointe->addTransition(this, &Case::ClickDroit, ModeNormal);

    ModeDecouvert->addTransition(this, &Case::ClickGaucheEtDroit, AppercuDesVoisins);

    AppercuDesVoisins->addTransition(this, &Case::PasDeClick, ModeVoisinsDecouverts);
    AppercuDesVoisins->addTransition(this, &Case::unPreview, ModeDecouvert);

    ModeVoisinsDecouverts->addTransition(this, &Case::reveal, ModeDecouvert);

    connect(ModeNormal, &QState::entered, [this]()
    {
        //this->setIcon(getIcon(IconeCase::Vide));
        this->setStyleSheet(NormalStyle);


    });

    connect(ModeAppercu, &QState::entered, [this]()
    {
        this->setStyleSheet(CaseCliqueStyle);
    });

    connect(AppercuDesVoisins, &QState::entered, [this]()
    {
        for (auto neighbor : _voisins)
            neighbor->preview();
    });

    connect(ModeVoisinsDecouverts, &QState::entered, [this]()
    {
        if (_NbDrapeauAutour == _NbMinesAutour && _NbMinesAutour)
            revealNeighbors();
        else
            unPreviewNeighbors();
        emit reveal();
    });

    connect(ModeDecouvert, &QState::entered, [this]()
    {
        unPreviewNeighbors();
        this->setChecked(true);
        if (!EstMine())
        {
            Afficher();

            if (!AdesMinesAutour())
                revealNeighbors();

            emit revealed();
        }
        else
        {
            emit Explosion();
            this->setStyleSheet(CaseCliqueStyle);
            QPushButton::setText("X");
           // setIcon(getIcon(IconeCase::Mine));
        }
    });

    connect(ModeDrapeauPointe, &QState::entered, [this]()
    {
        QPushButton::setStyleSheet(NormalStyle.arg("blue"));
         QPushButton::setText("?");
        for (auto neighbor : _voisins)
            neighbor->incrementNbDrapeauAutour();
        emit PointDrapeau(_isMine);

    });

    connect(ModeDrapeauPointe, &QState::exited, [this]()
    {
         QPushButton::setText("");
        for (auto neighbor : _voisins)
            neighbor->decrementNbDrapeauAutour();
        emit SuppressionDrappeau(_isMine);
    });

    connect(ModeInactive, &QState::entered, [this]{});

    _machine.addState(ModeNormal);
    _machine.addState(ModeAppercu);
    _machine.addState(AppercuDesVoisins);
    _machine.addState(ModeDrapeauPointe);
    _machine.addState(ModeDecouvert);
    _machine.addState(ModeVoisinsDecouverts);
    _machine.addState(ModeInactive);

    _machine.setInitialState(ModeNormal);
    _machine.start();
}

/*!
    *  \brief Gesstion de l'affichage des Indices
    *
    *  Elle permet d'afficher le nombre de Mines autour avec des couleurs différentes
    */
void Case::Afficher()
{
    QString color;
    switch (_NbMinesAutour)
    {
    case 1:
        color = "blue";
        break;
    case 2:
        color = "green";
        break;
    case 3:
        color = "red";
        break;
    case 4:
        color = "midnightblue";
        break;
    case 5:
        color = "maroon";
        break;
    case 6:
        color = "darkcyan";
        break;
    case 7:
        color = "black";
        break;
    case 8:
        color = "grey";
        break;
    default:
           break;
    }
    QPushButton::setStyleSheet(NumeroStyle.arg(color));
    if(_NbMinesAutour)
        QPushButton::setText(QString::number(_NbMinesAutour));
}

/*!
    *  \brief Corrige les Affichages à la fin de la partie
    */

void Case:: AfficheResultat(){
    if(EstMine()){
        QPushButton::setStyleSheet(NumeroStyle.arg("red"));
        QPushButton::setText("X");
    }
    if(EstMine() && AuneMarque()){
        QPushButton::setStyleSheet(NumeroStyle.arg("green"));
        QPushButton::setText("X");
    }
}
